features/filesystem/storagelite/source/StorageLite.h

```diff
@@ -0,0 +1,559 @@
+ * SPDX-License-Identifier: Apache-2.0
+ * Licensed under the Apache License, Version 2.0 (the License); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MBED_STORAGELITE_H
+#define MBED_STORAGELITE_H
+
+#if (STORAGELITE_ENABLED) || defined(DOXYGEN_ONLY)
+#include <stdint.h>
+#include <stdio.h>
+#include "platform/NonCopyable.h"
+#include "PlatformMutex.h"
+#include "BlockDevice.h"
+
+typedef enum {
+    STORAGELITE_SUCCESS                =  0,
+    STORAGELITE_READ_ERROR             = -1,
+    STORAGELITE_WRITE_ERROR            = -2,
+    STORAGELITE_NOT_FOUND              = -3,
+    STORAGELITE_DATA_CORRUPT           = -4,
+    STORAGELITE_BAD_VALUE              = -5,
+    STORAGELITE_BUFF_TOO_SMALL         = -6,
+    STORAGELITE_NO_SPACE_ON_BD         = -7,
+    STORAGELITE_OS_ERROR               = -8,
+    STORAGELITE_UNINITIALIZED          = -9,
+    STORAGELITE_MAX_FILES_REACHED      = -10,
+    STORAGELITE_NOT_SUPPORTED          = -11,
+    STORAGELITE_RB_PROTECT_ERROR       = -12,
+} storagelite_status_e;
+
+#ifndef STORAGELITE_MAX_FILES
+#define STORAGELITE_MAX_FILES 64
+#endif
+
+// defines 2 areas - active and nonactive, not configurable
+#define STORAGELITE_NUM_AREAS        2
+
+/** StorageLite class
+ *
+ *  Key value storage over a block device
+ */
+
+class StorageLite : mbed::NonCopyable<StorageLite> {
+public:
+
+    StorageLite();
+
+    virtual ~StorageLite();
+
+    /**
+     * @brief Returns maximal number of files.
+     *
+     * @returns Number of files.
+     */
+    size_t get_max_files() const;
+
+    /**
+     * @brief Returns file data, given file name.
+     *
+     * @param[in]  file_name            File name buffer (binary).
+     * @param[in]  file_name_size       File name buffer size.
+     * @param[in]  data_buf             Buffer to store data on.
+     * @param[in]  data_buf_size        Data buffer size (bytes).
+     * @param[out] actual_data_size     Actual data size (bytes).
+     *
+     * @returns STORAGELITE_SUCCESS           Success.
+     *          STORAGELITE_NOT_FOUND         File was not found.
+     *          STORAGELITE_READ_ERROR        Physical error reading data.
+     *          STORAGELITE_DATA_CORRUPT      Data is corrupt.
+     *          STORAGELITE_BAD_VALUE         Bad value in any of the parameters.
+     *          STORAGELITE_BUFF_TOO_SMALL    Not enough memory in user buffer.
+     */
+    int get(const uint8_t *file_name, size_t file_name_size, void *data_buf,
+            size_t data_buf_size, size_t& actual_data_size);
```

Comment:
Got to disagree here. This code is implemented in C++, so references should be used. Pointers fit to the C like programming scheme.

Reply:
We had a discussion about this a while back and agreed to use C pointers. I'm not sure where it is documented, @0xc0170?

For example, consider [line 709](https://github.com/ARMmbed/mbed-os/pull/6915/files/0a55174d36988ad2cfa9bbdab567400d8cf8711c#diff-340d6de032717e89f42100e4c0e8da0aR709):
``` cpp
os_ret = nvstore.set(header.nvstore_key, cmac_size, header.cmac);
```

Does this call modify header.cmac? What about cmac_size?

I understand your point about being consistent with other C++ code, however that has not been mbed's goal from what I understand.

Reply:
Well, I believe that the user should check the meaning of each parameter (in documentation for instance), where the parameter being modified or not is just one aspect. I personally find references very useful and dislike the usage of asterisks and ampersands throughout the function code and calls. My stand here is that you if you use C++, you can use any of its features apart from the ones strictly forbidden by code rules (exceptions, STL etc.). Otherwise you'd be just be programming in C.

Reply:
It's generally agreed upon that arrays in C/C++ aren't well integrated into the language. Though without references, you can still be confident that the address where `header.cmac` points won't change.

I do want to pull in @0xc0170 on this, at the end of the day consistency is more important than either of our opinions and @0xc0170 generally makes the final call on style for this reason.

Reply:
If consistency is what you seek, please observe too examples where references are used instead of pointers (and I guess there are more):
[LoRaWan get_plan function](https://github.com/ARMmbed/mbed-os/blob/e1a63f9046515df8d7f66ce251c3749812506bf6/features/lorawan/lorastack/mac/LoRaMacChannelPlan.h#L84)
(Note that in this case, the `set_plan` function uses the `const` directive, meaning that there's no way telling if it's a returning value or not from the function call, you just need to look at the header).
[NVstore get function](https://github.com/ARMmbed/mbed-os/blob/e1a63f9046515df8d7f66ce251c3749812506bf6/features/nvstore/source/nvstore.h#L127)
(just one example out of many in the NVStore code.